From 16fb2e2959184e57616265c1391da8fd9340c136 Mon Sep 17 00:00:00 2001
From: Paul Thomson <paulthomson@google.com>
Date: Thu, 16 Apr 2026 15:53:07 +0100
Subject: [PATCH 10/11] Fix 'KtDotQualifiedExpression.receiverExpression' for
 stubs

Sometimes, only one part of the 'KtDotQualifiedExpression'
is serialized to a stub. In cases when the receiver part is omitted,
'receiverExpression' worked incorrectly. In KT-81928 in particular, the
left part was an integer constant that previously was rejected by
'KtDotQualifiedExpression.getChildExpressionsByStub()'.

The commit removes the synthetic limitation for the children stubs
and also fixes the generic issue with `receiverExpression`.

The bug appeared in Standalone after its migration to the stub
infrastructure (in b402f60). The test
is implemented in the Standalone as usual Analysis API tests operate on
the PSI tree.

To avoid similar issues, we should build the test suite that would check
consistency of tree/stub PSI implementations (see KT-82189).

^KT-81928 Fixed

(Cherry picked from commit below, with some test changes dropped.)

https://github.com/JetBrains/kotlin/commit/9e1e6758f2d61563ded572f054ba4adb3d4c3225
https://youtrack.jetbrains.com/issue/KT-81928
Bug: 446964753
---
 .../kotlin/psi/KtDotQualifiedExpression.kt    | 30 +++++++++----------
 .../psi/stubs/elements/KtTokenSets.java       |  6 ----
 2 files changed, 15 insertions(+), 21 deletions(-)

diff --git a/compiler/psi/psi-api/src/org/jetbrains/kotlin/psi/KtDotQualifiedExpression.kt b/compiler/psi/psi-api/src/org/jetbrains/kotlin/psi/KtDotQualifiedExpression.kt
index 33c612f8f0c..d62fc598bf7 100644
--- a/compiler/psi/psi-api/src/org/jetbrains/kotlin/psi/KtDotQualifiedExpression.kt
+++ b/compiler/psi/psi-api/src/org/jetbrains/kotlin/psi/KtDotQualifiedExpression.kt
@@ -9,7 +9,6 @@ import com.intellij.lang.ASTNode
 import com.intellij.openapi.diagnostic.Logger
 import org.jetbrains.kotlin.KtStubBasedElementTypes
 import org.jetbrains.kotlin.psi.stubs.KotlinPlaceHolderStub
-import org.jetbrains.kotlin.psi.stubs.elements.KtTokenSets.INSIDE_DOT_QUALIFIED_EXPRESSION
 import org.jetbrains.kotlin.utils.exceptions.logErrorWithAttachment
 import org.jetbrains.kotlin.utils.exceptions.withPsiEntry
 
@@ -30,23 +29,25 @@ class KtDotQualifiedExpression : KtExpressionImplStub<KotlinPlaceHolderStub<KtDo
         get() = stubReceiverExpression ?: super.receiverExpressionOrNull
 
     private val stubReceiverExpression: KtExpression?
-        get() = stub?.let { stub ->
-            getChildExpressionsByStub(stub)?.getOrNull(0)
-        }
+        get() = getChildStubValue(index = 0)
 
     override val selectorExpression: KtExpression?
-        get() {
-            val stub = stub
-            if (stub != null) {
-                val childExpressionsByStub = getChildExpressionsByStub(stub)
-                if (childExpressionsByStub != null && childExpressionsByStub.size == 2) {
-                    return childExpressionsByStub[1]
-                }
+        get() = getChildStubValue(index = 1) ?: super.selectorExpression
+
+    private fun getChildStubValue(index: Int): KtExpression? {
+        val stub = stub
+        if (stub != null) {
+            val childExpressionsByStub = getChildExpressionsByStub(stub)
+            // If one part is missing in the stub, we cannot be sure whether we have a receiver or a selector
+            if (childExpressionsByStub != null && childExpressionsByStub.size == 2) {
+                return childExpressionsByStub[index]
             }
-            return super.selectorExpression
         }
 
-    private fun getChildExpressionsByStub(stub: KotlinPlaceHolderStub<KtDotQualifiedExpression>): Array<KtExpression>? {
+        return null
+    }
+
+    private fun getChildExpressionsByStub(stub: KotlinPlaceHolderStub<KtDotQualifiedExpression>): List<KtExpression>? {
         if (stub.getParentStubOfType(KtImportDirective::class.java) == null &&
             stub.getParentStubOfType(KtPackageDirective::class.java) == null &&
             stub.getParentStubOfType(KtValueArgument::class.java) == null
@@ -57,8 +58,7 @@ class KtDotQualifiedExpression : KtExpressionImplStub<KotlinPlaceHolderStub<KtDo
             )
             return null
         } else {
-            @OptIn(KtImplementationDetail::class)
-            val expressions = stub.getChildrenByType(INSIDE_DOT_QUALIFIED_EXPRESSION, KtExpression.ARRAY_FACTORY)
+            val expressions = stub.childrenStubs.mapNotNull { it.psi as? KtExpression }
             if (expressions.size !in 1..2) {
                 LOG.logErrorWithAttachment("Invalid stub structure. DOT_QUALIFIED_EXPRESSION must have one or two children. Was: ${expressions.size}") {
                     withPsiEntry("file", containingFile)
diff --git a/compiler/psi/psi-api/src/org/jetbrains/kotlin/psi/stubs/elements/KtTokenSets.java b/compiler/psi/psi-api/src/org/jetbrains/kotlin/psi/stubs/elements/KtTokenSets.java
index 15aaf68fd64..90c9b2fb607 100644
--- a/compiler/psi/psi-api/src/org/jetbrains/kotlin/psi/stubs/elements/KtTokenSets.java
+++ b/compiler/psi/psi-api/src/org/jetbrains/kotlin/psi/stubs/elements/KtTokenSets.java
@@ -21,12 +21,6 @@ public interface KtTokenSets {
 
     TokenSet INSIDE_DIRECTIVE_EXPRESSIONS = TokenSet.create(REFERENCE_EXPRESSION, DOT_QUALIFIED_EXPRESSION);
 
-    /**
-     * TokenSet containing all expression types that can be a child of [KtDotQualifiedExpression].
-     */
-    @KtImplementationDetail
-    TokenSet INSIDE_DOT_QUALIFIED_EXPRESSION = TokenSet.create(REFERENCE_EXPRESSION, CALL_EXPRESSION, DOT_QUALIFIED_EXPRESSION);
-
     // typeArguments? valueArguments : typeArguments : arrayAccess
     TokenSet POSTFIX_OPERATIONS = TokenSet.create(KtTokens.PLUSPLUS, KtTokens.MINUSMINUS, KtTokens.EXCLEXCL, KtTokens.DOT, KtTokens.SAFE_ACCESS);
 
-- 
2.54.0.rc2.544.gc7ae2d5bb8-goog

